### PR TITLE
Editor: Display post locked dialog when other user is editing

### LINKED ISSUE #1613: Editor: posts do not lock when someone else is editing
Raised by @gibrown 

Barring collaborative editing, we should probably lock the post when multiple people try to edit it.

WP Admin does this:

screen shot 2015-12-15 at tue dec 15 10 11 48 am

I assume if I edited from an existing page while someone else is doing so then bad things will happen (I haven't tried).

### PR DESCRIPTION
**Pull request by aduth**
Closes #1613

__Very WIP:__ The dialog is currently shown, but design, behaviors to assign the meta and take over a post, plus clean-up and tests, are yet to be completed.

This pull request seeks to display a dialog when attempting to edit a post which is being edited by another user. They are prompted to choose between returning to the all posts screen or take over editing controls of the post.

WIP

_[Reference Design](https://cloud.githubusercontent.com/assets/4389/12146259/a06112fc-b48a-11e5-98cc-073896e8fdea.png)_

__Technical Questions:__

I've not yet settled on a technical implementation for "taking over" a post. In Core, this is achieved through the [Heartbeat API](https://wptavern.com/how-to-take-control-of-the-wordpress-heartbeat-api), refreshing the post lock and verifying that another user has not taken over in the process of doing so. Because we have no such heartbeat API in Calypso, we must consider an alternative. The two obvious options to me are (a) polling and (b) websockets, which the latter being preferable though more complex.

__Testing instructions:__

_TBD_

### CODE REVIEW

**Comment by dmsnell** [comment]
We could use Pinghub here and beyond "Take control" we could actually "Request to take control" and wait for a release of control from the currently-editing author. As I understand it, Pinghub channels are broadcast so the pattern might look like this:

 - Customer routes to editor for given post/page
 - Calypso connects to Pinghub channel `/editorLock/siteId/postId`
 - All other attached clients announce if they are editing the post
 - If we need control, we issue request
 - Whichever client has control sees request and responds with yes/no, times out after given delay (5s or so?)

Pinghub is already available through `wpcom.js`

**Comment by aduth** [comment]
@dmsnell This would work great if it were a greenfield project. Since a locking feature exists using post meta (the `_edit_lock` flag), ideally we could have backwards compatibility using it in case not all users on a site use Calypso. Then again, maybe this mismatch is a very rare case and it'd be fine enough to optimize for the best overall solution.

File: client/post-editor/editor-post-lock-dialog/index.jsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+					},
+					{
+						action: 'confirm',
+						label: translate( 'Take Over', { context: 'verb' } ),
+						onClick: this.takeOver
+					}
+				] }
+				onClose={ this.returnToPosts }
+				isVisible>
+				Post Being Edited
```
Comment (dmsnell) [inline client/post-editor/editor-post-lock-dialog/index.jsx]: This needs translation here, or a post title, or something.

**Comment by dmsnell** [comment]
@aduth do you have a status on this? seems like a reasonable PR to push through. what is WIP about it?

**Comment by aduth** [comment]
This is very much needed, yes. It is work-in-progress in that:

- There's no design to it. See proposed design at https://github.com/Automattic/wp-calypso/issues/1613#issuecomment-169359004 . Getting information about the user currently editing seems like it should be fairly reasonable (`<QueryUsers />`).
- Selectors should be documented, include tests, and moved to `client/state/selectors`
- The big one: "Take over" needs a technical direction, as noted in "Technical questions" of original comment and your own suggestion of Pinghub. I think there should be some backwards compatibility with the post meta even if we went the route of Websockets. Otherwise we might consider some polling to update the post meta / retrieve the latest post meta relating to post locking.

**Comment by dmsnell** [comment]
> suggestion of Pinghub. I think there should be some backwards compatibility with the post meta even if we went the route of Websockets. Otherwise we might consider some polling to update the post meta / retrieve the latest post meta relating to post locking.

probably as a first step I think polling would be simpler and faster-to-ship. thanks for the updates

**Comment by alisterscott** [comment]
Closing as this PR is stale - please reopen if necessary

**Pull request closed without merging**